File: android/app/src/main/java/dev/pam/nativeapp/modules/StorageModule.kt
package dev.pam.nativeapp.modules

import android.content.Context
import dev.pam.nativeapp.protocol.WireMap
import dev.pam.nativeapp.protocol.WireValue
import java.util.concurrent.Executors
import java.util.concurrent.RejectedExecutionException
import java.util.concurrent.atomic.AtomicBoolean

internal class StorageModule(context: Context) : NativeModule, AutoCloseable {
    private val preferences = context.getSharedPreferences("pam-native", Context.MODE_PRIVATE)
    private val closed = AtomicBoolean()
    private val executor = Executors.newSingleThreadExecutor { runnable ->
        Thread(runnable, "pam-storage").apply { isDaemon = true }
    }

    override fun invoke(
        method: String,
        payload: ByteArray,
        completion: ModuleCompletion,
    ) {
        if (closed.get()) {
            completion.complete(ModuleResultStatus.FAILURE, "Storage module is closed".toByteArray())
            return
        }
        try {
            executor.execute {
                runCatching {
                    val values = WireMap.decode(payload)
                    val key = (values["key"] as? WireValue.Text)?.value
                        ?: error("Storage key is required")
                    require(key.matches(Regex("[A-Za-z0-9_.-]{1,128}"))) { "Invalid storage key" }
                    when (method) {
                        "get" -> {
                            val value = preferences.getString(key, null)
                            WireMap.encode(
                                value?.let { mapOf("value" to WireValue.Text(it)) } ?: emptyMap(),
                            )
                        }
                        "set" -> {
                            val value = (values["value"] as? WireValue.Text)?.value
                                ?: error("Storage value is required")
                            require(value.toByteArray().size <= MAX_STORAGE_VALUE_BYTES) {
                                "Storage value exceeds 256 KiB"
                            }
                            check(preferences.edit().putString(key, value).commit()) {
                                "Storage write failed"
                            }
                            WireMap.encode(emptyMap())
                        }
                        else -> error("Unknown storage method")
                    }
                }.fold(
                    onSuccess = { completion.complete(ModuleResultStatus.SUCCESS, it) },
                    onFailure = {
                        completion.complete(
                            ModuleResultStatus.FAILURE,
                            (it.message ?: "Storage operation failed").toByteArray(),
                        )
                    },
                )
            }
        } catch (_: RejectedExecutionException) {
            completion.complete(ModuleResultStatus.FAILURE, "Storage module is closed".toByteArray())
        }
    }

    override fun close() {
        if (!closed.compareAndSet(false, true)) return
        executor.shutdownNow()
    }

    private companion object {
        const val MAX_STORAGE_VALUE_BYTES = 256 * 1024
    }
}

File: ios/Sources/PamNative/Modules/StorageModule.swift
import Foundation

public final class StorageModule: NativeModule, ClosableNativeModule, @unchecked Sendable {
    private let defaults = UserDefaults.standard
    private let queue = DispatchQueue(label: "pam.native.storage", qos: .utility)
    private var closed = false

    public init() {}

    public func invoke(method: String, payload: Data, completion: @escaping ModuleCompletion) {
        if closed {
            completion(.failure, "Storage module is closed".data(using: .utf8) ?? Data())
            return
        }

        queue.async {
            do {
                let values = try WireMap.decode(payload)
                guard case let .text(key)? = values["key"],
                      self.isValidKey(key) else {
                    throw RuntimeError("Storage key is required")
                }

                switch method {
                case "get":
                    if let value = self.defaults.string(forKey: key) {
                        completion(.success, try WireMap.encode(["value": .text(value)]))
                    } else {
                        completion(.success, Data())
                    }
                case "set":
                    guard case let .text(value)? = values["value"] else {
                        throw RuntimeError("Storage value is required")
                    }
                    let bytes = value.data(using: .utf8) ?? Data()
                    guard bytes.count <= 256 * 1024 else {
                        throw RuntimeError("Storage value exceeds 256 KiB")
                    }
                    self.defaults.set(value, forKey: key)
                    completion(.success, Data())
                default:
                    completion(.failure, "Unknown storage method".data(using: .utf8) ?? Data())
                }
            } catch {
                completion(.failure, (error.localizedDescription).data(using: .utf8) ?? Data())
            }
        }
    }

    public func close() {
        closed = true
    }

    private func isValidKey(_ key: String) -> Bool {
        guard (1...128).contains(key.count) else { return false }
        return key.allSatisfy { char in
            char.isASCII && (char.isLetter || char.isNumber || char == "_" || char == "-" || char == ".")
        }
    }

    private struct RuntimeError: LocalizedError {
        let message: String
        init(_ value: String) { self.message = value }
        var errorDescription: String? { message }
    }
}
